omnimove_external_control/src/omnimove_external_control.cpp

```diff
@@ -0,0 +1,357 @@
+  }
+  //   RCLCPP_INFO(rclcpp::get_logger("OmnimoveExternalControl"), "Am waiting to read something if
+  //   available");
+  size_t bytes_received = client_socket_->read_some(boost::asio::buffer(buffer));
+  // RCLCPP_INFO(rclcpp::get_logger("OmnimoveExternalControl"), "received %lu", bytes_received);
+
+  read_buffer_.insert(read_buffer_.end(), buffer.begin(), buffer.begin() + bytes_received);
+  // RCLCPP_INFO(rclcpp::get_logger("OmnimoveExternalControl"), "Inserted into buffer");
+
+  // parse all messages from the buffer.
+  // incomplete messages need to be kept for further reading.
+  ExternalControlData readData = parseLastMessageFromBuffer();
+  //  RCLCPP_INFO(rclcpp::get_logger("OmnimoveExternalControl"), "Finished parsing into buffer");
+
+  if (readData.isDataValid())
+  {
+    if (agv_type_ == "caterpillar")
+    {
+      velocity_state_[0] = (double)readData.speedX();
+      velocity_state_[1] = (double)readData.speedW();
+      position_state_[0] = (double)readData.posPillar1();
+      velocity_state_[2] = (double)readData.speedPillar1();
+      position_state_[1] = (double)readData.posPillar2();
+      velocity_state_[3] = (double)readData.speedPillar2();
+      position_state_[2] = (double)readData.posPillar3();
+      velocity_state_[4] = (double)readData.speedPillar3();
+      position_state_[3] = (double)readData.posPillar4();
+      velocity_state_[5] = (double)readData.speedPillar4();
+      position_state_[4] = (double)readData.posShield();
+      velocity_state_[6] = (double)readData.speedShield();
+    }
+    else
+    {
+      velocity_state_[0] = (double)readData.speedX();
+      velocity_state_[1] = (double)readData.speedY();
+      velocity_state_[2] = (double)readData.speedW();
+    }
+  }
+  return return_type::OK;
+}
+
+hardware_interface::return_type OmnimoveExternalControl::write(
+  const rclcpp::Time &, const rclcpp::Duration &)
+{
+  //  RCLCPP_INFO_STREAM(rclcpp::get_logger("OmnimoveExternalControl"), "writing "<<
+  //  velocity_commands_[0]
+  //      <<" "<<velocity_commands_[1]<<" "<<velocity_commands_[2]);
+
+  if (agv_type_ == "caterpillar")
+  {
+    if (velocity_commands_ != last_sent_velocity_commands_)
+    {
+      last_sent_velocity_commands_ = velocity_commands_;
+      last_sent_velocity_time_ = boost::chrono::system_clock::now();
+      client_socket_->send(ExternalControlCaterpillarDriveCommand(
+                             velocity_commands_[0], velocity_commands_[1], position_commands_[0],
+                             position_commands_[1], position_commands_[2], position_commands_[3],
+                             position_commands_[4])
+                             .getSerialisedData());
+    }
+    else
+    {
+      if (
+        (boost::chrono::system_clock::now() - last_sent_velocity_time_) >
+        boost::chrono::milliseconds(vel_cmd_timeout_ms_))
+      {
+        client_socket_->send(ExternalControlStopCommand().getSerialisedData());
+      }
```

Comment:
this block is duplicated a few lines later, I think with a little restructuring these two could be combined outside of the check for the agv type